server/src/main/java/org/elasticsearch/cluster/metadata/InternalDesiredNodesMembershipService.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class InternalDesiredNodesMembershipService implements ClusterStateListener, DesiredNodesMembershipService {
+    private final ClusterService clusterService;
+    private final Set<DesiredNode> members;
+    private String latestHistoryId = null;
+
+    public InternalDesiredNodesMembershipService(ClusterService clusterService) {
+        this.clusterService = clusterService;
+        this.members = new HashSet<>();
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public synchronized void clusterChanged(ClusterChangedEvent event) {
+        final var clusterState = event.state();
+        final var desiredNodes = DesiredNodes.latestFromClusterState(clusterState);
+        if (desiredNodes == null) {
+            members.clear();
+            return;
+        }
+
+        if (event.nodesChanged()) {
+            final var nodesDelta = event.nodesDelta();
+
+            for (DiscoveryNode addedNode : nodesDelta.addedNodes()) {
+                final var desiredNode = desiredNodes.find(addedNode.getExternalId());
+                if (desiredNode != null) {
+                    members.add(desiredNode);
+                }
+            }
+        } else if (event.changedCustomMetadataSet().contains(DesiredNodesMetadata.TYPE) || latestHistoryId == null) {
+            if (desiredNodes.historyID().equals(latestHistoryId) == false) {
+                members.clear();
+            }
+            latestHistoryId = desiredNodes.historyID();
+
+            final Set<DesiredNode> removedDesiredNodes = new HashSet<>(members);
+            desiredNodes.nodes().forEach(removedDesiredNodes::remove);
```

Comment:
I think we may want to refine this to work on just the external id? The idea being that a node that was a member in the prior desired nodes is also a member with the new desired nodes even if it is temporarily out of the cluster while applying this. We can leave that for later I think.